staging/kos/pkg/registry/network/subnet/strategy.go

```diff
@@ -145,135 +126,83 @@ func (*subnetStrategy) AllowUnconditionalUpdate() bool {
-				thisSubnet.Namespace,
-				thisSubnet.Name,
-				aSubnet.Namespace,
-				aSubnet.Name)
-			errs = append(errs, field.Forbidden(field.NewPath("metadata", "namespace"), errMsg))
-		}
+	// Retrieve the summaries of all the other existing subnets with the same
+	// VNI so that we can check for conflicts.
+	potentialRivals, err := ss.subnetsWithVNI(fmt.Sprint(subnetSummary.VNI))
+	if err != nil {
+		return append(errs, field.InternalError(nil, err))
 	}
 
-	return errs
-}
-
-func (s *subnetStrategy) checkCIDRDoesNotOverlap(thisSubnet *network.Subnet, allSubnetsWithSameVNI []*network.Subnet) field.ErrorList {
-	errs := field.ErrorList{}
-
-	// No need to check if parse was successful, as the check has been already
-	// preformed in method checkCIDRFormat.
-	thisSubnetParsed := parseSubnet(thisSubnet)
-
-	for _, aSubnet := range allSubnetsWithSameVNI {
-		if thisSubnet.UID == aSubnet.UID {
-			// Do not compare the subnet under validation against itself (needed
-			// in case of updates).
+	// Check whether there are Namespace and CIDR conflicts with other subnets.
+	for _, pr := range potentialRivals {
+		if subnetSummary.SameSubnetAs(pr) {
+			// Do not compare this subnet against itself.
 			continue
 		}
-		// No need to check if parse was successuful, as aSubnet has already
-		// been created. If it couldn't be parsed its creation would have
-		// failed.
-		aSubnetParsed := parseSubnet(aSubnet)
-		if thisSubnetParsed.overlaps(aSubnetParsed) {
-			errMsg := fmt.Sprintf("subnet %s/%s IPv4 range (%s) overlaps with that of subnet %s/%s (%s). IPv4 ranges "+
-				"for subnets with the same VNI MUST be disjoint",
-				thisSubnet.Namespace,
-				thisSubnet.Name,
-				thisSubnet.Spec.IPv4,
-				aSubnet.Namespace,
-				aSubnet.Name,
-				aSubnet.Spec.IPv4)
-			errs = append(errs, field.Forbidden(field.NewPath("spec", "ipv4"), errMsg))
+		glog.V(2).Infof("Validating %s against %s", subnetSummary.NamespacedName, pr.NamespacedName)
+		if subnetSummary.NSConflict(pr) {
+			errs = append(errs, field.Forbidden(field.NewPath("spec", "vni"), fmt.Sprintf("subnets with same VNI must be within same namespace, but %s has the same VNI and a different namespace", pr.NamespacedName)))
+		}
+		if !malformedCIDR && subnetSummary.CIDRConflict(pr) {
+			errs = append(errs, field.Forbidden(field.NewPath("spec", "ipv4"), fmt.Sprintf("subnets with same VNI must have disjoint CIDRs, but CIDR overlaps with %s's", pr.NamespacedName)))
 		}
 	}
 
 	return errs
 }
 
-type parsedSubnet struct {
-	baseU, lastU uint32
-}
-
-func parseSubnet(subnet *network.Subnet) (ps parsedSubnet) {
-	_, ipNet, err := net.ParseCIDR(subnet.Spec.IPv4)
+func (ss *subnetStrategy) subnetsWithVNI(vni string) ([]*subnet.Summary, error) {
+	subnets, err := ss.subnetIndexer.ByIndex(subnetVNIIndex, vni)
 	if err != nil {
-		return
+		return nil, fmt.Errorf(".ByIndex failed for index %s and vni %s: %s", subnetVNIIndex, vni, err.Error())
+	}
+	glog.V(3).Infof("Found %d subnets with vni %s", len(subnets), vni)
```

Comment:
I think this is a rather small detail and potentially large log message, V level 3 is too prominent; I think something like 5 or 6 is more appropriate.